api/search_test.go

```diff
@@ -184,6 +184,49 @@ func TestSearchHandlerFunc(t *testing.T) {
 		c.So(esMock.MultiSearchCalls(), c.ShouldHaveLength, 0)
 	})
 
+	c.Convey("Should return BadRequest for invalid datasetID params", t, func() {
+		qbMock := newQueryBuilderMock(nil, nil)
+		esMock := newDpElasticSearcherMock(nil, nil)
+		trMock := newResponseTransformerMock(nil, nil)
+
+		searchHandler := SearchHandlerFunc(validator, qbMock, &config.Config{}, &ClientList{DpESClient: esMock}, trMock)
+
+		req := httptest.NewRequest("GET", "http://localhost:8080/search?dataset_id=q", http.NoBody)
+		resp := httptest.NewRecorder()
+
+		searchHandler.ServeHTTP(resp, req)
+		c.So(resp.Code, c.ShouldEqual, http.StatusBadRequest)
+		c.So(resp.Body.String(), c.ShouldContainSubstring, "Invalid DatasetID(s): ")
+		c.So(qbMock.BuildSearchQueryCalls(), c.ShouldHaveLength, 0)
+		c.So(esMock.MultiSearchCalls(), c.ShouldHaveLength, 0)
+	})
+
+	c.Convey("Should return OK for valid datasetID parameter", t, func() {
+		searchBytes, _ := json.Marshal(searches)
+		qbMock := newQueryBuilderMock(searchBytes, nil)
+		esMock := newDpElasticSearcherMock([]byte(validESResponse), nil)
+		trMock := newResponseTransformerMock([]byte(validTransformedResponse), nil)
+
+		searchHandler := SearchHandlerFunc(validator, qbMock, &config.Config{}, &ClientList{DpESClient: esMock}, trMock)
+
+		req := httptest.NewRequest("GET", baseURL+validQueryParam+"&dataset_id=QNA", http.NoBody)
+		resp := httptest.NewRecorder()
+
+		searchHandler.ServeHTTP(resp, req)
+
+		c.So(resp.Code, c.ShouldEqual, http.StatusOK)
+		c.So(resp.Body.String(), c.ShouldResemble, validTransformedResponse)
+		c.So(qbMock.BuildSearchQueryCalls(), c.ShouldHaveLength, 1)
+		c.So(qbMock.BuildSearchQueryCalls()[0].Req.Term, c.ShouldResemble, validQueryParam)
+		c.So(esMock.MultiSearchCalls(), c.ShouldHaveLength, 1)
+		actualRequest := string(esMock.MultiSearchCalls()[0].Searches[0].Query)
+		c.So(actualRequest, c.ShouldResemble, expectedQuery)
+		c.So(trMock.TransformSearchResponseCalls(), c.ShouldHaveLength, 1)
+		c.So(trMock.TransformSearchResponseCalls()[0].Highlight, c.ShouldBeTrue)
+		actualResponse := string(trMock.TransformSearchResponseCalls()[0].ResponseData)
+		c.So(actualResponse, c.ShouldResemble, validESResponse)
```

Comment:
I don't think you've checked that it's actually passing the dataset_id to the query?

Reply:
Ah maybe it's further down.